### ISSUE TITLE
Display the next few characters on the client before synchronisation to improve visual latency

### ISSUE THREAD

**Issue by parrenin** [CollaboraOnline/online#1308]
**Is your feature request related to a problem? Please describe.**

There is a latency when you enter text in any of the components (document, spreadsheet, presentation).

**Describe the solution you'd like**

My suggestion to improve the visual latency when entering text would be to display the next few characters using javascript on the client before there are synchronized to the server and you get the tile update.
One could use a different color (e.g., grey instead of black text) to outline these characters are not yet synchronized.
I think it could improve the comfort when entering text.

**Describe alternatives you've considered**

**Additional context**

Maybe you have considered this feature already and discarded it for good reasons.
In this case, I would be interested to know the reasons.

**Comment by parrenin** [comment]
A precision on what I have in mind: I do not expect you to write a full layout engine in javascript.
The important thing here is just to have an instantaneous visual feedback on the characters you are currently entering.
For example, at the end of a line, you can just draw the next few characters in the margin, no need to take into account the line layout.

**Comment by mmeeks** [comment]
It would be interesting to see the latency you're experiencing. If you go to help->about and tripple click, you should have some tools show that which can help with that.  You could take a video with the console up showing what's going on - and the console should have helpful timestamps from keystroke to the text showing up.

Naturally it's possible to be much cleverer in future about caching things - but I'd love to tweak & tune what we have now (which is simpler) to be far more efficient first; would be great to see what use-case is particularly slow for you, and what your ping times are and what mode you use COOL in.

Thanks !

**Comment by parrenin** [comment]
Thanks for the tip.
So the console indicates an average ping time of ~320 ms and worst time of ~1 s.
I am on your EU  demo server, since my employer has not (yet?) opted for Collabora Online and I have not yet installed CODE on my own server.
I did not do an extensive testing of the software, just trying to modify an academic article a few pages long with figures.
I think the ping time is the limiting factor here.
~320 ms is quite good already, but you do not feel like it is instantaneous when you write text, hence my suggestion.

**Comment by mmeeks** [comment]
Wow, 320ms is a terrible ping time. Where are you based ? https://www.cloudping.info/ suggests this is a reasonable ping to South Korea from Ireland; my ping to Ireland is 20ms - but I'd expect something well under 100ms can you run ping on the console? Thanks.

**Comment by parrenin** [comment]
I am based in Grenoble, France and work from home right now.
cloudping gives a ping of 52 ms to Ireland.
If I run 'ping demo.eu.collaboraonline.com' on the console, I get 30-50 ms.
Strange that the console in Collabora Online gives a larger ping time.
I checked again a now it is around 200 ms.

**Comment by mmeeks** [comment]
Hmm - indeed, that is interesting - there is a lot of debug code there, quite possibly it's measuring something different; mine says best 9ms, avg 26ms, worst 51ms - which seems overly fast. What are your three numbers ? (at the bottom left in the overlay) ? hmm - my numbers in the files say 5 seconds or so ;-) I guess that's a bug - quite possibly that needs updating, possibly if we don't send a tile response since it is unchanged we mis-account that debug data. Interested in hacking on that ? code is in loleaflet/src/layer/tile/TileLayer.js see _debugSetTimes ono. Thanks =)

**Comment by mmeeks** [comment]
What type of COOL server are you using ? are you using an appimage one-click install (which simulates a websocket cunningly), or a custom bare-metal, or Docker install ? thanks.

**Comment by parrenin** [comment]
Right now I am just testing your demo.eu.collaboraonline.com server.

**Comment by parrenin** [comment]
Now it is best: 32ms, avg: 418 ms, worst: 793 ms.

**Comment by rpear** [comment]
@mmeeks, I agree with @parrenin that this would be a great enhancement to be considered. I live in Australia and the Internet is poor in most places.  I appreciate the enormous latency improvements that have been introduced with COnline in the last year or so, the request from @parrenin is appealing and sounds logical after remaining tile improvements have been completed.  I think I appreciate the complexity of this request, guessing where to put the text, the font, the size, refreshing all the overwritten tiles, etc. Should this ticket be tagged confirmed even if it might not happen? 

(Internet is poor in Australia thanks to [Tony Abbott](https://www.computerworld.com/article/3466613/in-his-own-words-tony-abbott-on-the-nbn.html), an Australian prime minister in 2013, a buddy of Rupert Murdoch - Sky and cable so they didn't want fast internet in AU and trashed the fibre migration between them. In 2023 the Murdoch's put Tony Abbott on the Fox board as a $500,000 per year presumably as a reward for helping trash the AU internet and for other assistance with the trashy misinformation network.)

**Comment by rpear** [comment]
I am testing it on Online 24.04.14.1 and it feels very responsive even with my ping times at 300 - 1000ms. My ping times are usually 300 - 1000ms due to my location in AU on 4G, the server is in the UK (https://share.collabora.com/).

It used to feel a bit laggy on previous times, but I haven't noticed annoying lag for many weeks perhaps even months now. It could be luck, or the current server load or something else. I'll make a mental note to update this ticket if I notice it is laggy again.

From Cloudping.info:
IBM Cloud
us-south (Dallas)	208 ms
us-east (Washington DC)	235 ms
eu-gb (London)	335 ms
eu-de (Frankfurt)	339 ms
au-syd (Sydney)	44 ms
